### PR TITLE
Multi catalog

### PR DESCRIPTION
**Pull request by SzymonPobiega**
Implements multi-catalog using `table@[schema]@[catalog]` address format that is backwards-compatible with pre-3 and 3.0.0 transports. 

TODO:
 * [x] ~Add wire-compatibility tests to `EndToEnd`~ Adding tests is handled by separate issue #353 
 * [x] Decide if we want to support new catalog part in LegacyMultiInstanceMode - we don't
 * [ ] Discuss validation of multi-instance connection string

### CODE REVIEW

**Comment by SzymonPobiega** [comment]
@MarcinHoppe @HEskandari please take a look. I removed the assertion from the SQL injection via endpoint reply address test because I could not make it work.

**Comment by SzymonPobiega** [comment]
Also I am not sure how to test the wire-level compat with 2.X and 3.0 with regards to the address form. Should I copy the 3.0 address parting code to a test and verify it works against addresses containing the catalog name?

**Comment by MarcinHoppe** [comment]
Maybe we could add tests to cover multi-* addressing here:

https://github.com/Particular/EndToEnd/tree/master/src/TransportCompatibilityTests/SqlServerV3

?

**Comment by SzymonPobiega** [comment]
@MarcinHoppe sounds like a plan. But we need to at least have a 3.1 package on myget to get these tests to work, right?

**Comment by MarcinHoppe** [comment]
AFAIK yes.

File: src/NServiceBus.SqlServer/Addressing/QueueAddress.cs [hunk 1]
```diff
@@ -1,54 +1,135 @@
...
     using System.Data.SqlClient;
+    using System.Linq;
 
     class QueueAddress
     {
-        public QueueAddress(string tableName, string schemaName)
+        public QueueAddress(string catalog, string schemaName, string tableName)
         {
             Guard.AgainstNullAndEmpty(nameof(tableName), tableName);
+            if (catalog == "")
```
Comment (HEskandari) [inline src/NServiceBus.SqlServer/Addressing/QueueAddress.cs]: String.Empty? Should we also handle case of a whitespace only string?

File: src/NServiceBus.SqlServer.UnitTests/QueueAddressTests.cs [hunk 2]
```diff
@@ -7,32 +7,39 @@
...
-        [TestCase("my.table@[my.schema]", "my.table", "my.schema")]
-        public void Should_parse_address(string transportAddress, string expectedTableName, string expectedSchema)
+        [TestCase("table@schema", "table", "schema", null)]
+        [TestCase("my.table@my.schema", "my.table", "my.schema", null)]
+        [TestCase("table", "table", null, null)]
+        [TestCase("table@[my.schema]", "table", "my.schema", null)]
+        [TestCase("my.table@[my.schema]", "my.table", "my.schema", null)]
+        [TestCase("my.table@[my.schema]@[my.catalog]", "my.table", "my.schema", "my.catalog")]
+        [TestCase("my.table@my.schema@my.catalog", "my.table", "my.schema", "my.catalog")]
+        [TestCase("my.table@[]@[my.catalog]", "my.table", "", "my.catalog")]
```
Comment (HEskandari) [inline src/NServiceBus.SqlServer.UnitTests/QueueAddressTests.cs]: I would expected this format to fail (just my opinion though).

Comment (SzymonPobiega) [inline src/NServiceBus.SqlServer.UnitTests/QueueAddressTests.cs]: The last one is the way to say: send to `my.table` in `my.catalog` using default configured schema. We can't get rid of `[]` (or just space) because it would be interpreted as `my.table` in `my.catalog` schema in the current catalog.

**Comment by SzymonPobiega** [comment]
@Particular/sqlserver-transport-maintainers ping.

**Comment by SzymonPobiega** [comment]
@Particular/sqlserver-transport-maintainers ping :-)

**Comment by HEskandari** [comment]
@SzymonPobiega Is there anything else left here? I thought it was already reviewed?

**Comment by SzymonPobiega** [comment]
@HEskandari I was waiting for the final `LGTM`s from you and @MarcinHoppe :-)

**Comment by MarcinHoppe** [comment]
On it now @SzymonPobiega...

File: src/NServiceBus.SqlServer/SqlServerTransportSettingsExtensions.cs [hunk 3]
```diff
@@ -62,6 +62,25 @@ public static TransportExtensions<SqlServerTransport> UseSchemaForQueue(this Tra
         }
 
         /// <summary>
+        /// Specifies custom schema for given endpoint.
+        /// </summary>
+        /// <param name="transportExtensions">The <see cref="TransportExtensions{T}" /> to extend.</param>
+        /// <param name="endpointName">Endpoint name.</param>
+        /// <param name="catalog">Custom catalog value.</param>
+        /// <returns></returns>
+        public static TransportExtensions<SqlServerTransport> UseCatalogForEndpoint(this TransportExtensions<SqlServerTransport> transportExtensions, string endpointName, string catalog)
```
Comment (MarcinHoppe) [inline src/NServiceBus.SqlServer/SqlServerTransportSettingsExtensions.cs]: Do we also need to specify it for queues and not endpoints (e.g. audit, error queue)?

File: src/NServiceBus.SqlServer/SqlServerTransportInfrastructure.cs [hunk 4]
```diff
@@ -155,15 +156,38 @@ public override TransportSubscriptionInfrastructure ConfigureSubscriptionInfrast
...
+                if (parser.TryGetValue("Initial Catalog", out catalog) || parser.TryGetValue("Database", out catalog))
+                {
+                    result = result.SetProperty(SettingsKeys.CatalogPropertyKey, (string) catalog);
+                }
+                else
+                {
+                    throw new Exception("Initial Catalog property is mandatory in the connection string.");
+                }
+            }
+            else if (!settings.HasSetting("PublicReturnAddress"))
```
Comment (MarcinHoppe) [inline src/NServiceBus.SqlServer/SqlServerTransportInfrastructure.cs]: Is this condition related? If not, should go into a separate issue and PR.

File: src/NServiceBus.SqlServer/Addressing/QueueAddress.cs [hunk 5]
```diff
@@ -1,54 +1,147 @@
...
 {
+    using System;
+    using System.Collections.Generic;
     using System.Data.SqlClient;
+    using System.Linq;
 
     class QueueAddress
     {
-        public QueueAddress(string tableName, string schemaName)
+        public QueueAddress(string catalog, string schemaName, string tableName)
```
Comment (MarcinHoppe) [inline src/NServiceBus.SqlServer/Addressing/QueueAddress.cs]: `catalogName` would be more consistent with `schemaName` and `tableName`?

File: src/NServiceBus.SqlServer/Addressing/QueueAddress.cs [hunk 6]
```diff
@@ -1,54 +1,147 @@
...
         {
             Guard.AgainstNullAndEmpty(nameof(tableName), tableName);
+            if (catalog != null && IsEmptyOrWhiteSpace(catalog))
```
Comment (MarcinHoppe) [inline src/NServiceBus.SqlServer/Addressing/QueueAddress.cs]: Can this be shortened to a call to `String.IsNullOrWhiteSpace`?

File: src/NServiceBus.SqlServer/Addressing/QueueAddress.cs [hunk 7]
```diff
@@ -1,54 +1,147 @@
...
+            {
+                throw new ArgumentException("Catalog name cannot be empty or whitespace.", nameof(catalog));
+            }
+            if (catalog != null && schemaName == null)
+            {
+                throw new Exception("If catalog is specified, schema name has to be specified too.");
```
Comment (MarcinHoppe) [inline src/NServiceBus.SqlServer/Addressing/QueueAddress.cs]: Why? Can't we assume a default user schema for a given catalog?

Comment (MarcinHoppe) [inline src/NServiceBus.SqlServer/Addressing/QueueAddress.cs]: Or even `dbo`?

File: src/NServiceBus.SqlServer/Addressing/QueueAddress.cs [hunk 8]
```diff
@@ -1,54 +1,147 @@
...
         }
 
+        public string Catalog { get; }
         public string TableName { get; }
         public string SchemaName { get; }
+        public string Unquoted { get; }
+        public string Quoted { get; }
+        public bool HasDefinedSchema => !string.IsNullOrWhiteSpace(SchemaName);
 
         public static QueueAddress Parse(string address)
```
Comment (MarcinHoppe) [inline src/NServiceBus.SqlServer/Addressing/QueueAddress.cs]: I know this might sound like adding another problem but... regular expressions?

File: src/NServiceBus.SqlServer.AcceptanceTests/When_ReplyTo_address_does_not_exist.cs [hunk 9]
```diff
@@ -25,7 +25,6 @@ await Scenario.Define<Context>()
             var failedMessage = exception.FailedMessages.Single();
 
             Assert.That(failedMessage.Exception.Message, Contains.Substring("Failed to send message to"));
-            Assert.That(failedMessage.Exception.Message, Contains.Substring("error]] VALUES(NEWID(), NULL, NULL, 1, NULL, '', NULL); DROP TABLE [Victim]]; INSERT INTO [error"));
```
Comment (MarcinHoppe) [inline src/NServiceBus.SqlServer.AcceptanceTests/When_ReplyTo_address_does_not_exist.cs]: What's the reason for removing this line?

File: src/NServiceBus.SqlServer.AcceptanceTests/ConfigureEndpointSqlServerTransport.cs [hunk 10]
```diff
@@ -2,6 +2,7 @@
 using System.Collections.Generic;
 using System.Data.SqlClient;
 using System.Linq;
+using System.Runtime.InteropServices;
```
Comment (MarcinHoppe) [inline src/NServiceBus.SqlServer.AcceptanceTests/ConfigureEndpointSqlServerTransport.cs]: The change does not seem to use anything from this namespace?

File: src/NServiceBus.SqlServer.AcceptanceTests/MultiCatalog/When_using_multicatalog.cs [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+        public class Receiver : EndpointConfigurationBuilder
+        {
+            public Receiver()
+            {
+                EndpointSetup<DefaultServer>(c =>
+                {
+                    c.UseTransport<SqlServerTransport>()
+                        .ConnectionString(ReceiverConnectionString)
+                        .UseSchemaForEndpoint(ReceiverEndpoint, "receiver")
+                        .UseCatalogForEndpoint(SenderEndpoint, "nservicebus1");
```
Comment (MarcinHoppe) [inline src/NServiceBus.SqlServer.AcceptanceTests/MultiCatalog/When_using_multicatalog.cs]: Why does the receiver has to know the sender catalog? Shouldn't it come from the return address?

File: src/NServiceBus.SqlServer.AcceptanceTests/When_ReplyTo_address_does_not_exist.cs (hunk 9 above)
Comment (SzymonPobiega) [inline src/NServiceBus.SqlServer.AcceptanceTests/When_ReplyTo_address_does_not_exist.cs]: I left you a comment here https://github.com/Particular/NServiceBus.SqlServer/pull/330#issuecomment-269943264. Basically I could not figure out what should be the expression that simulates the injection attack now that we have better address encapsulation (the quotes are no longer hard coded in the SQL template).

File: src/NServiceBus.SqlServer.AcceptanceTests/MultiCatalog/When_using_multicatalog.cs (hunk 11 above)
Comment (SzymonPobiega) [inline src/NServiceBus.SqlServer.AcceptanceTests/MultiCatalog/When_using_multicatalog.cs]: Removed.

File: src/NServiceBus.SqlServer/Addressing/QueueAddress.cs (hunk 6 above)
Comment (SzymonPobiega) [inline src/NServiceBus.SqlServer/Addressing/QueueAddress.cs]: Nope. This is an equivalent of `IsNotNullAnd(EmptyOrWhiteSpace)` :( Meaning, it can be null, but if it is not null, then it cannot be whitespace.

**Comment by SzymonPobiega** [comment]
@MarcinHoppe while trying to apply your suggestions I got a bit stuck. Will come back to this after my vacation.

File: src/NServiceBus.SqlServer/Addressing/QueueAddress.cs [hunk 12]
```diff
@@ -1,54 +1,147 @@
...
+            }
+            if (catalog != null && schemaName == null)
+            {
+                throw new Exception("If catalog is specified, schema name has to be specified too.");
+            }
+            Catalog = Unquote(catalog);
+            TableName = tableName;
+            SchemaName = Unquote(schemaName);
+            Quoted = string.Join(".", GetAllParts().Select(QuoteIdentifier));
+            Unquoted = string.Join(".", GetAllParts());
```
Comment (tmasternak) [inline src/NServiceBus.SqlServer/Addressing/QueueAddress.cs]: Does it really belong here? It's used for table creation lock and logs. The first could be moved to QueueCreator (the logic of removing unsupported characters). Do you see any problems with having quoted identifiers in log statements?

File: src/NServiceBus.SqlServer/Addressing/QueueAddress.cs [hunk 13]
```diff
@@ -1,54 +1,147 @@
...
+        }
 
-            TableName = UnescapeIdentifier(tableName);
-            SchemaName = UnescapeIdentifier(schemaName);
+        static bool IsEmptyOrWhiteSpace(string value)
```
Comment (tmasternak) [inline src/NServiceBus.SqlServer/Addressing/QueueAddress.cs]: Why not Linq and inlining in the if?

File: src/NServiceBus.SqlServer/Addressing/QueueAddress.cs [hunk 14]
```diff
@@ -1,54 +1,147 @@
...
+        public string Unquoted { get; }
+        public string Quoted { get; }
+        public bool HasDefinedSchema => !string.IsNullOrWhiteSpace(SchemaName);
 
         public static QueueAddress Parse(string address)
         {
-            if (address.Contains("@"))
+            var location = address;
+
+            var rightBrackets = 0;
```
Comment (tmasternak) [inline src/NServiceBus.SqlServer/Addressing/QueueAddress.cs]: Do we need to count brackets? How about we scan the string from right-to-left and have a bool flat that tracks if we are in inside `[]`. If yes than we ignore any `@` otherwise we treat is as part separator?

File: src/NServiceBus.SqlServer/Addressing/QueueAddress.cs [hunk 15]
```diff
@@ -1,54 +1,147 @@
 ﻿namespace NServiceBus.Transport.SQLServer
 {
+    using System;
+    using System.Collections.Generic;
     using System.Data.SqlClient;
+    using System.Linq;
 
     class QueueAddress
```
Comment (tmasternak) [inline src/NServiceBus.SqlServer/Addressing/QueueAddress.cs]: Do we lack a concept here? Should our types reflect two cases? Something like
 1. A result of address parsing that might not be fully specified This is the current `QueueAddress`.
 2. A fully specified address. Maybe `CanonicalAddress` that has all three parts specified.

In such case we could have a well defined areas of code that expect `CanonicalAddress` and  a `Parser` that would translate from `QueueAddress` to `CanonicalAddress` using configuration and defaults.

File: src/NServiceBus.SqlServer/SqlServerTransportInfrastructure.cs [hunk 16]
```diff
@@ -155,15 +156,38 @@ public override TransportSubscriptionInfrastructure ConfigureSubscriptionInfrast
...
             string schema;
             if (schemaSettings.TryGet(instance.Endpoint, out schema) == false)
             {
                 schema = addressParser.DefaultSchema;
             }
 
-            return instance.SetProperty(SettingsKeys.SchemaPropertyKey, schema);
+            var result = instance.SetProperty(SettingsKeys.SchemaPropertyKey, schema);
+
+            if (!settings.HasSetting(SettingsKeys.ConnectionFactoryOverride) && !settings.HasSetting(SettingsKeys.LegacyMultiInstanceConnectionFactory))
```
Comment (tmasternak) [inline src/NServiceBus.SqlServer/SqlServerTransportInfrastructure.cs]: Do we need the `LegacyMultiInstanceConnectionFactory` check here. `SqlServerTransportInfrastructure` will not be used in Legacy Mode.

File: src/NServiceBus.SqlServer/SqlServerTransportSettingsExtensions.cs (hunk 3 above)
Comment (tmasternak) [inline src/NServiceBus.SqlServer/SqlServerTransportSettingsExtensions.cs]: I think we need to do it. Otherwise there would be no way to specify custom catalog for those queues.

**Comment by SzymonPobiega** [comment]
@Particular/sqlserver-transport-maintainers  I've updated the implementation. All tests should be green now.

**Comment by SzymonPobiega** [comment]
@tmasternak can you review https://github.com/Particular/NServiceBus.SqlServer/pull/353 ? This is I think the last requirement to finish this one.

**Comment by SzymonPobiega** [comment]
@MarcinHoppe @MarcinHoppe I've redesigned this a bit. Please review when you have time.

File: src/NServiceBus.SqlServer/SqlServerTransport.cs [hunk 17]
```diff
@@ -45,14 +36,48 @@ bool LegacyMultiInstanceModeTurnedOn(SettingsHolder settings)
...
+            else
+            {
+                var catalog = GetDefaultCatalog(settings, connectionString);
+                var addressParser = new QueueAddressTranslator((string)catalog, "dbo", defaultSchemaOverride, queueSchemaSettings);
+                return new SqlServerTransportInfrastructure(addressParser, settings, connectionString);
+            }
+        }
 
-            return new SqlServerTransportInfrastructure(addressParser, settings, connectionString);
+        static object GetDefaultCatalog(SettingsHolder settings, string connectionString)
```
Comment (MarcinHoppe) [inline src/NServiceBus.SqlServer/SqlServerTransport.cs]: Why not cast object to string before returning from this method? This way we bother the caller.

File: src/NServiceBus.SqlServer/Legacy/MultiInstance/LegacySqlConnectionFactory.cs [hunk 18]
```diff
@@ -34,7 +32,8 @@ void ValidateConnectionPool(string transportAddress, string connectionString)
                 Logger.Warn(validationResult.Message);
             }
 
-            SetValidated(transportAddress);
+            //TODO: is this correct?
+            //SetValidated(transportAddress);
```
Comment (MarcinHoppe) [inline src/NServiceBus.SqlServer/Legacy/MultiInstance/LegacySqlConnectionFactory.cs]: If this is removed then the call to `HasValidated` should also be removed.

File: src/NServiceBus.SqlServer/SqlServerTransportSettingsExtensions.cs [hunk 19]
```diff
@@ -54,9 +54,47 @@ public static TransportExtensions<SqlServerTransport> UseSchemaForQueue(this Tra
...
+        /// <summary>
+        /// Specifies custom schema for given queue.
+        /// </summary>
+        /// <param name="transportExtensions">The <see cref="TransportExtensions{T}" /> to extend.</param>
+        /// <param name="queueName">Queue name.</param>
+        /// <param name="catalog">Custom catalog value.</param>
+        /// <returns></returns>
+        public static TransportExtensions<SqlServerTransport> UseCatalogForQueue(this TransportExtensions<SqlServerTransport> transportExtensions, string queueName, string catalog)
+        {
+            Guard.AgainstNullAndEmpty(nameof(queueName), queueName);
```
Comment (SimonCropp) [inline src/NServiceBus.SqlServer/SqlServerTransportSettingsExtensions.cs]: minor but we should guard against null on `transportExtensions`

File: src/NServiceBus.SqlServer/Legacy/MultiInstance/LegacySqlConnectionFactory.cs (hunk 18 above)
Comment (SzymonPobiega) [inline src/NServiceBus.SqlServer/Legacy/MultiInstance/LegacySqlConnectionFactory.cs]: I've fixed it using concurrent dictionary

File: src/NServiceBus.SqlServer/SqlServerTransportSettingsExtensions.cs (hunk 19 above)
Comment (SzymonPobiega) [inline src/NServiceBus.SqlServer/SqlServerTransportSettingsExtensions.cs]: Added more guards :-)

**Comment by SzymonPobiega** [comment]
Rebased against develop and fixed the issues. Ready for another round of reviews.

**Comment by SzymonPobiega** [comment]
I've added the compat tests for the multi-catalog. @tmasternak @MarcinHoppe please re-review this one

**Comment by SzymonPobiega** [comment]
Merged together with native timeouts.
